rfcs/021-data_science_in_the_pipeline/README.md

```diff
@@ -0,0 +1,54 @@
+# RFC 021: Data science in the pipeline
+
+**Status:** 🏗 Draft 🚧
+
+**Last updated:** 2020/03/06
+
+## Motivation
+
+We want to augment works and images with data inferred from them using data science techniques: for example, feature vectors and colour palettes for images.
+
+Currently, we do this by holding both some form of index (usually a set of points in a vector space) as well as a model in a separate service - for example, https://labs.wellcomecollection.org/feature-similarity. This has significant drawbacks: data is duplicated, outdated, patchy, and perhaps most importantly, must be searched as a wholly separate all-or-nothing operation outside of our ES indices.
+
+To resolve this, we want to bring data science services into the pipeline, and use them to augment works/images with data that can be indexed and searched in ES. This RFC details how that might look.
+
+## Proposal
+
+We can write down some desiderata for any proposed solution:
+
+- Data science (DS) logic can be written in Python
+- DS services do not know about our data model(s)
+- Pipeline services do not know about data science
+- DS services do not know about SQS, message passing, the Actor model, etc
+- DS models are persistent and separate to DS inferrers
+- DS models can be retrained on demand
+
+These suggest 3 distinct, but loosely coupled, services:
+
+- **Inferrer**: A Python service that provides a synchronous API (most likely RESTful HTTP) that consumes whatever is needed to infer the new data, which it outputs.
+- **Model Trainer**: A Python service that can consume records from the catalogue index in bulk in order to train a model, and outputs/stores a persistent representation of this model for the *inferrer* to use.
+- **Inference Manager**: A Scala service that lives in the existing pipeline and contains "the usual" Wellcome message-passing, Akka, etc logic & libraries, which sychronously calls the `inferrer` and attaches the new data to the work/image (by populating a field) before passing it along.
+
+The usage of these services would look like this:
+
+![An architecture diagram for data science services](https://user-images.githubusercontent.com/4429247/76087593-631cc700-5fae-11ea-819f-3545e6dc7c65.png)
```

Comment:
Where does this fit into the existing pipeline? What are things & augmented things?

Reply:
I had previously imagined that augmented data would be fed into the pipeline as another source, matched & merged as other data sources are.

Reply:
An inference task definition could live anywhere in the pipeline after transformation. A thing has optional fields that contain data that the inference task definition might populate, after which the thing is augmented. I'll clarify that in the RFC. 

I don't really see the advantages of having augmented data as a new source - it would introduce a cycle into the pipeline (in that the source would need to have a source of existing works) and the merger would just be "undoing" this cycle.

We want the data science stuff to be source-agnostic, merely adding extra searchable information to works - in the absence of something like change feeds in ES it seems to me like that should live in the pipeline.

Reply:
Before the ingestor perhaps?

Reply:
Yeah, that's where I imagined it. Makes sense for doing this for images as well.